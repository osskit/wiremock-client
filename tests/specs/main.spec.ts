import fetch from 'node-fetch';
import { createMapping, waitForCalls, hasMadeCalls, HttpMethod, LogLevel, reset, create, Call } from '../../dist';

interface Body {
  field1: string;
  numberField: number;
  array: { field2: string }[];
}

const body = { field1: 'value1', numberField: 1234, array: [{ field2: 'value2' }] };
const firstBody = { field1: 'value1', numberField: 1234, array: [{ field2: 'someValue' }] };
const secondBody = { field1: 'value1', numberField: 12345, array: [{ field2: 'someValue' }] };

describe('tests', () => {
  beforeEach(async () => {
    create();
    await reset();
  });

  describe('create', () => {
    it('creates a new client with different configuration ', async () => {
      create({ baseUrl: 'http://localhost:9090', logLevel: LogLevel.Warn, continueOnFailure: false });
      const request = await createMapping({ request: { urlPathPattern: '/someUrl', method: HttpMethod.Post } });

      await fetch('http://localhost:9090/someUrl', { method: HttpMethod.Post, body: JSON.stringify(body) });
      expect(await hasMadeCalls(request)).toBeTruthy();
    });
  });

  describe('waitForCalls', () => {
    it('returns calls information for post request', async () => {
      const request = await createMapping({ request: { urlPathPattern: '/someUrl', method: HttpMethod.Post } });

      await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Post,
        body: JSON.stringify(body),
        headers: { 'Content-Type': 'application/json' },
      });
      await fetch('http://localhost:8080/someUrl', { method: HttpMethod.Post, body: JSON.stringify(body) });

      const calls = await waitForCalls(request);

      expect(calls).toHaveLength(2);
      expect(calls).toMatchSnapshot([{ loggedDate: expect.any(Number) }, { loggedDate: expect.any(Number) }]);
    });

    it('returns calls information for get request', async () => {
      const request = await createMapping({ request: { urlPathPattern: '/someUrl', method: HttpMethod.Get } });

      await fetch('http://localhost:8080/someUrl?someQueryParam=someValue', {
        method: HttpMethod.Get,
        headers: { 'Content-Type': 'application/json' },
      });

      const calls = await waitForCalls(request);

      expect(calls).toHaveLength(1);
      expect(calls).toMatchSnapshot([{ loggedDate: expect.any(Number) }]);
    });

    it('returns calls information for different request method to the same url', async () => {
      const putRequest = await createMapping({ request: { urlPathPattern: '/someUrl', method: HttpMethod.Put } });
      const postRequest = await createMapping({ request: { urlPathPattern: '/someUrl', method: HttpMethod.Post } });

      await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Post,
        body: JSON.stringify(body),
        headers: { 'Content-Type': 'application/json' },
      });

      await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Put,
        body: JSON.stringify(body),
        headers: { 'Content-Type': 'application/json' },
      });

      const putCalls = await waitForCalls(putRequest);

      expect(putCalls).toHaveLength(1);
      expect(putCalls).toMatchSnapshot([{ loggedDate: expect.any(Number) }]);

      const postCalls = await waitForCalls(postRequest);

      expect(postCalls).toHaveLength(1);
      expect(putCalls).toMatchSnapshot([{ loggedDate: expect.any(Number) }]);
    });

    it('returns ordered calls information', async () => {
      const request = await createMapping({ request: { urlPathPattern: '/someUrl', method: HttpMethod.Post } });

      await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Post,
        body: JSON.stringify(secondBody),
        headers: { 'Content-Type': 'application/json' },
      });

      await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Post,
        body: JSON.stringify(firstBody),
        headers: { 'Content-Type': 'application/json' },
      });

      const calls = await waitForCalls(request, {
        orderBy: (a: Call, b: Call) => (a.body as Body).numberField - (b.body as Body).numberField,
      });

      expect(calls).toMatchSnapshot([{ loggedDate: expect.any(Number) }, { loggedDate: expect.any(Number) }]);
    });
  });

  describe('createMapping', () => {
    it('returns mocked response as string on string body pattern', async () => {
      await createMapping({
        request: { urlPathPattern: '/someUrl', method: HttpMethod.Post },
        response: { status: 200, body: 'some body once told me' },
      });

      const response = await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Post,
        body: JSON.stringify(body),
        headers: { 'Content-Type': 'application/json' },
      });

      expect(await response.text()).toMatchSnapshot();
    });

    it('returns mocked response as string on base64Body pattern', async () => {
      await createMapping({
        request: { urlPathPattern: '/someUrl', method: HttpMethod.Post },
        response: { status: 200, body: 'c29tZSBib2R5IG9uY2UgdG9sZCBtZQ==' },
      });

      const response = await fetch('http://localhost:8080/someUrl', {
        method: HttpMethod.Post,
        body: JSON.stringify(body),
        headers: { 'Content-Type': 'application/json' },
      });

      expect(await response.text()).toMatchSnapshot();
    });
  });

  describe('hasMadeCalls', () => {
    it('returns true if calls were made', async () => {
      const request = await createMapping({
        request: { urlPathPattern: '/someUrl', method: HttpMethod.Put },
        response: { status: 204, jsonBody: [{ returnValue: 'someReturnValue' }] },
      });

      await fetch('http://localhost:8080/someUrl', { method: HttpMethod.Put, body: JSON.stringify(body) });

      expect(await hasMadeCalls(request)).toBeTruthy();
    });

    it('returns false if calls were not made', async () => {
      const request = await createMapping({
        request: { urlPathPattern: '/someUrl', method: HttpMethod.Post },
        response: { status: 204, jsonBody: [{ returnValue: 'someReturnValue' }] },
      });

      expect(await hasMadeCalls(request)).toBeFalsy();
    });
  });
});
