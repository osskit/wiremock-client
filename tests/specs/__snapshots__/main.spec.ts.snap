// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests createMapping returns mocked response as string on base64Body pattern 1`] = `"c29tZSBib2R5IG9uY2UgdG9sZCBtZQ=="`;

exports[`tests createMapping returns mocked response as string on string body pattern 1`] = `"some body once told me"`;

exports[`tests waitForCalls returns bodyAsString 1`] = `
[
  {
    "body": "yay",
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "3",
      "Content-Type": "text/plain;charset=UTF-8",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "POST",
    "queryParams": {},
    "url": "/someUrl",
  },
]
`;

exports[`tests waitForCalls returns calls information for different request method to the same url 1`] = `
[
  {
    "body": {
      "array": [
        {
          "field2": "value2",
        },
      ],
      "field1": "value1",
      "numberField": 1234,
    },
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "68",
      "Content-Type": "application/json",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "PUT",
    "queryParams": {},
    "url": "/someUrl",
  },
]
`;

exports[`tests waitForCalls returns calls information for different request method to the same url 2`] = `
[
  {
    "body": {
      "array": [
        {
          "field2": "value2",
        },
      ],
      "field1": "value1",
      "numberField": 1234,
    },
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "68",
      "Content-Type": "application/json",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "PUT",
    "queryParams": {},
    "url": "/someUrl",
  },
]
`;

exports[`tests waitForCalls returns calls information for get request 1`] = `
[
  {
    "body": "",
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "GET",
    "queryParams": {
      "someQueryParam": {
        "key": "someQueryParam",
        "values": [
          "someValue",
        ],
      },
    },
    "url": "/someUrl?someQueryParam=someValue",
  },
]
`;

exports[`tests waitForCalls returns calls information for post request 1`] = `
[
  {
    "body": {
      "array": [
        {
          "field2": "value2",
        },
      ],
      "field1": "value1",
      "numberField": 1234,
    },
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "68",
      "Content-Type": "application/json",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "POST",
    "queryParams": {},
    "url": "/someUrl",
  },
  {
    "body": {
      "array": [
        {
          "field2": "value2",
        },
      ],
      "field1": "value1",
      "numberField": 1234,
    },
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "68",
      "Content-Type": "text/plain;charset=UTF-8",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "POST",
    "queryParams": {},
    "url": "/someUrl",
  },
]
`;

exports[`tests waitForCalls returns ordered calls information 1`] = `
[
  {
    "body": {
      "array": [
        {
          "field2": "someValue",
        },
      ],
      "field1": "value1",
      "numberField": 1234,
    },
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "71",
      "Content-Type": "application/json",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "POST",
    "queryParams": {},
    "url": "/someUrl",
  },
  {
    "body": {
      "array": [
        {
          "field2": "someValue",
        },
      ],
      "field1": "value1",
      "numberField": 12345,
    },
    "headers": {
      "Accept": "*/*",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "*",
      "Connection": "keep-alive",
      "Content-Length": "72",
      "Content-Type": "application/json",
      "Host": "localhost:8080",
      "User-Agent": "undici",
      "sec-fetch-mode": "cors",
    },
    "loggedDate": Any<Number>,
    "method": "POST",
    "queryParams": {},
    "url": "/someUrl",
  },
]
`;
